File: Methods/Java/Methods/Extra/Steffensen.java
package Methods.Extra;

import Share.Functions.IFunction;

public class Steffensen {

    private final Aikten aikten = new Aikten();

    public double execute(IFunction f, double tolerance, int maxIterations, double approximation) {
        double x0, x1, x2, x3;

        x0 = approximation;
        x1 = f.evaluate(x0);
        x2 = f.evaluate(x1);
        x3 = aikten.acceleration(x0, x1, x2);

        for (int i = 1; i <= maxIterations && aikten.absoluteError(x0, x3) > tolerance; ++i) {
            x0 = x3;
            x1 = f.evaluate(x0);
            x2 = f.evaluate(x1);
            x3 = aikten.acceleration(x0, x1, x2);
        }
        return x3;
    }
}

File: Methods/Python/Extra/Steffensen.py
import Aikten as ak
import Error as e

def execute(f, tolerance, maxIterations, approximation):


    x0 = f(1)
    x1 = f(2)
    x2 = f(3)
    x3 = ak.acceleration(x0, x1, x2)

    i = 1
    while i <= (maxIterations and e.absoluteError(x0, x3) > tolerance):
        x0 = x3
        x1 = f(x0)
        x2 = f(x1)
        x3 = ak.acceleration(x0, x1, x2)
        i += 1
    return x3